src/kaleidoscope_internal/storage.h

```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+#define _INLINE_STORAGE_HANDLER_FOR_PLUGIN(PLUGIN)                                       __NL__ \
+  EventHandler__::call(PLUGIN, current_offset, hook_args...);                            __NL__ \
+  current_offset += PLUGIN.storageSize();
+
+#define _INLINE_STORAGE_OFFSET_COUNTER(PLUGIN)                                           __NL__ \
+  if ((uint16_t)&PLUGIN == (uint16_t)&plugin)                                            __NL__ \
+    return current_offset;                                                               __NL__ \
+  current_offset += PLUGIN.storageSize();
+
+#define __KALEIDOSCOPE_INIT_STORAGE_PLUGINS(...)                                         __NL__ \
+  /* Iterate through plugins, calling onStorageReport for each of them. */               __NL__ \
+  template<typename EventHandler__, typename... Args__ >                                 __NL__ \
+  static kaleidoscope::EventHandlerResult                                                __NL__ \
+  apply_storage(Args__&&... hook_args) {                                                 __NL__ \
+    uint16_t current_offset = 0;                                                         __NL__ \
+    MAP(_INLINE_STORAGE_HANDLER_FOR_PLUGIN, __VA_ARGS__)                                 __NL__ \
+    return kaleidoscope::EventHandlerResult::OK;                                         __NL__ \
+  }                                                                                      __NL__ \
+                                                                                         __NL__ \
+  template<typename EventHandler__, typename Plugin__>                                   __NL__ \
+  static uint16_t apply_offsetCount(Plugin__ &plugin) {                                  __NL__ \
+    uint16_t current_offset = 0;                                                         __NL__ \
+    MAP(_INLINE_STORAGE_OFFSET_COUNTER, __VA_ARGS__)                                     __NL__ \
+    return current_offset;                                                               __NL__ \
+  }
+
+#define _KALEIDOSCOPE_INIT_STORAGE_HANDLERS()                                            __NL__ \
+  namespace kaleidoscope_internal {                                                      __NL__ \
+  struct EventHandler_onStorageReport {                                                  __NL__ \
+    template<typename Plugin__>                                                          __NL__ \
+    static kaleidoscope::EventHandlerResult                                              __NL__ \
+    call(Plugin__ &plugin, uint16_t offset) {                                            __NL__ \
+      if (plugin.storageSize()) {                                                        __NL__ \
+        Serial.print(plugin.name());                                                     __NL__ \
```

Comment:
Are these Serial.prints just for debugging while you're doing development?

Reply:
No, this is the `onStorageReport` "hook" itself. Plugins themselves do not implement this method, they implement `storageSize` and `name`, which this one uses.

When we want to dump the storage layout, this is the hook to call, hence the Serial.prints.

Yeah, I'm not convinced either that this is the best way. But it was the simplest.

Reply:
I don’t think we ought to be explicitly calling Serial in the core... Happy to chat about this latet

Reply:
Ack, lets chat.

Reply:
Perhaps instead of calling Serial.print in core, we could be passing the data to a function inside Focus?

Reply:
That's a possibility, yeah.